packages/effect/src/Effect.ts

```diff
@@ -2159,6 +2159,61 @@ export const uninterruptibleMask: <A, E, R>(
   f: (restore: <AX, EX, RX>(effect: Effect<AX, EX, RX>) => Effect<AX, EX, RX>) => Effect<A, E, R>
 ) => Effect<A, E, R> = core.uninterruptibleMask
 
+// -------------------------------------------------------------------------------------
+// lifting
+// -------------------------------------------------------------------------------------
+
+/**
+ * Transforms a `Predicate` function into an `Effect` returning the input value if the predicate returns `true`
+ * or failing with specified error if the predicate fails
+ *
+ * @param predicate - A `Predicate` function that takes in a value of type `A` and returns a boolean.
+ *
+ * @example
+ * import { pipe, Effect } from "effect"
+ *
+ * const isPositive = (n: number): boolean => n > 0
+ *
+ * assert.deepStrictEqual(
+ *   pipe(
+ *     1,
+ *     Effect.liftPredicate(isPositive, n => `${n} is not positive`)
+ *   ),
+ *   Effect.succeed(1)
+ * )
+ * assert.deepStrictEqual(
+ *   pipe(
+ *     0,
+ *     Effect.liftPredicate(isPositive, n => `${n} is not positive`)
+ *   ),
+ *   Effect.fail("0 is not positive")
+ * )
+ *
+ * @category lifting
+ * @since 3.4.0
+ */
+export const liftPredicate: {
+  <A, B extends A, E>(refinement: Refinement<NoInfer<A>, B>, orFailWith: (a: NoInfer<A>) => E): (a: A) => Effect<B, E>
+  <A, E>(
+    predicate: Predicate<NoInfer<A>>,
+    orFailWith: (a: NoInfer<A>) => E
+  ): (a: A) => Effect<A, E>
+  <A, E, B extends A>(
+    self: A,
+    refinement: Refinement<A, B>,
+    orFailWith: (a: A) => E
+  ): Effect<B, E>
+  <A, E>(
+    self: A,
+    predicate: Predicate<NoInfer<A>>,
+    orFailWith: (a: NoInfer<A>) => E
+  ): Effect<A, E>
+} = dual(
+  3,
+  <A, E>(a: A, predicate: Predicate<A>, orFailWith: (a: A) => E): Effect<A, E> =>
+    predicate(a) ? succeed(a) : fail(orFailWith(a))
```

Comment:
This needs to be wrapped in `Effect.suspend`, in case the predicate throws an error.